data store: wipe workflow data on reload.

oliver-sanders opened the pull request:
The UI side of https://github.com/cylc/cylc-flow/pull/5715

In combination with https://github.com/cylc/cylc-flow/pull/5715, this closes #1480

* When a workflow is reloaded or restarted (node restart infers reload), there may be objects left behind in the store which have been wiped out in the scheduler by the configuration change.
* To handle this, we must wipe all objects stored on the workflow (e.g. TaskProxies, TaskDefs, Jobs, etc), and rebuild from scratch.
* Note, we don't need to wipe the data on the Workflow object itself (e.g. port, stateTotals, etc), these things will be refreshed by this or subsequent deltas.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

MetRonnie commented on src/views/SimpleTree.vue:
```diff
@@ -160,6 +166,13 @@ fragment PrunedDelta on Pruned {
 
+# We must always request "id" and "reloaded" for the workflow.
+# These are required by the data store.
+fragment workflowData on Workflow {
```
```suggestion
fragment WorkflowData on Workflow {
```

MetRonnie commented on CHANGES.md:
Conflicting

MetRonnie reviewed (changes requested):
Something is going wrong here when you open the graph view at the same time as the tree view. Both stop updating.

oliver-sanders commented on src/services/treeCallback.js:
```diff
@@ -25,27 +25,41 @@ class CylcTreeCallback extends DeltasCallback {
+    // do this, we can end up with nodes in the store which aren't meant to be
+    // there and won't get pruned.
+    if (deltas.updated?.workflow?.reloaded) {
+      store.commit('workflows/REMOVE_CHILDREN', (deltas.updated.workflow.id))
+    }
+    if (deltas.added?.workflow?.reloaded) {
+      store.commit('workflows/REMOVE_CHILDREN', (deltas.added.workflow.id))
```
To test, try jamming in `console.log` statements or whatever here to ensure that the `REMOVE_CHILDREN` store method is being called.

oliver-sanders commented on src/views/README.md:
```diff
@@ -25,3 +25,61 @@ E.G. the Tree *view* displays Cylc cycles/tasks/jobs in a collapsible hierarchy.
 the Workflows *view*.
+
+
+## Subscriptions
```
Added some details here as the rules for subscription writing are getting a little complex.

See also https://github.com/cylc/cylc-ui/issues/862 for context, and a possible way to simplify things for view writers.

oliver-sanders commented:
Started this again from scratch:

* I've moved the subscriptions out of the common module into the views to which they pertain.
* The idea of the subscription merging interface is that each view only requests the fields it needs so code reuse is not desirable here.
* I've been through the subscriptions and pulled out a number of fields which weren't required by the views which were requesting them.
* I've also added a couple of things in which the views weren't requesting but should have been.
* The tree and table now have different subscriptions containing only the things they need rather than the superset of the two.
* I've added the `reloaded` field for each of the workflow query views (i.e. the views which get their data from the "workflow" subscription). The other subscriptions don't need this field:
  * Application subscription is used for `Workflows`, `Dashboard` & `WorkflowsTable`, since these only request information about the workflow, not about tasks, etc, they don't need to be wiped on reload.
  * Log subscriptions are reload agnostic.
  * The analysis view uses queries not subscriptions so doesn't require this logic.

The tree and graph seem to be playing together nicely.

MetRonnie reviewed (commented):
Seems to work well. Spotted a couple of minor issues after a reload which may or may not be specific to this branch

1. Succeeded tasks without jobs (I have a feeling we came across this before)
  
    ![image](https://github.com/cylc/cylc-ui/assets/61982285/f39ec8a9-4bc4-4d2a-89dc-4c4ecd9a5d6b)

2. The Graph View arrow placement went a bit errant for a moment and I got this error and couldn't pan or zoom anymore

    ![image](https://github.com/cylc/cylc-ui/assets/61982285/08718af6-0d5e-42e2-99d9-8c04e1a6ca72)

oliver-sanders commented on src/views/Dashboard.vue:
```diff
@@ -145,7 +145,47 @@ import subscriptionComponentMixin from '@/mixins/subscriptionComponent'
+  }
+}
+
+fragment Deltas on Deltas {
```
It's required for subscription merging to work.

MetRonnie replied:
Is subscription merging done by a library or is it in-house?

oliver-sanders replied:
It's in house code that uses a simple system where it treats a GraphQL document as a nested set (which is pretty accurate) and merges like items. So all subscriptions must match the same format to merge correctly. The merged query is logged to the console for debugging purposes.

oliver-sanders commented on src/views/Table.vue:
```diff
@@ -33,7 +33,126 @@ import subscriptionComponentMixin from '@/mixins/subscriptionComponent'
+  cyclePoints: familyProxies (ids: ["*/root"]) {
+    ...CyclePointData
+  }
+  familyProxies {
+    ...FamilyProxyData
```
Shouldn't be, though it was before.

oliver-sanders commented:
> Succeeded tasks without jobs (I have a feeling we came across this before)

I haven't seen that one in testing, though I have seen https://github.com/cylc/cylc-flow/issues/5215

If you're able to reproduce, could you check whether the job info was sent through in the deltas?

oliver-sanders commented:
I can't reproduce either reported issue?

oliver-sanders commented:
Turns out that repeatably reloading the workflow with this branch is a really good way of triggering the firstParent bug.

Tried testing with https://github.com/cylc/cylc-ui/pull/1513 merged in and everything is looking good after 25 reloads:

![Screenshot from 2023-10-16 16-58-34](https://github.com/cylc/cylc-ui/assets/16705946/bd3dd37d-e877-4286-9662-56e1be9f3e6c)

```diff
diff --git a/src/services/treeCallback.js b/src/services/treeCallback.js
index 570c3ac4..1ab9308c 100644
--- a/src/services/treeCallback.js
+++ b/src/services/treeCallback.js
@@ -38,9 +38,11 @@ class CylcTreeCallback extends DeltasCallback {
     // do this, we can end up with nodes in the store which aren't meant to be
     // there and won't get pruned.
     if (deltas.updated?.workflow?.reloaded) {
+      console.warn('REMOVE_CHILDREN')
       store.commit('workflows/REMOVE_CHILDREN', (deltas.updated.workflow.id))
     }
     if (deltas.added?.workflow?.reloaded) {
+      console.warn('REMOVE_CHILDREN')
       store.commit('workflows/REMOVE_CHILDREN', (deltas.added.workflow.id))
     }
   }

```

oliver-sanders commented:
@MetRonnie 

* Applied suggestions
* Removed delta fragments from all queries

MetRonnie reviewed (commented):
I've tested out again, and the only problem I can see is the jobs disappear on reload (includes running and succeeded jobs) (seems to be at least `latestJob` goes missing?). This does not happen on master / #1513 presumably because the data was not wiped.

hjoliver reviewed (approved):
LGTM, tested as working 👍
